### PR TITLE
Wire max-shape inference into workspace estimation

### PR DESCRIPTION
**Pull request by chilo-ms**
### Description

This PR wires estimation-only input-shape hints into the two-level workspace-estimation flow described in #29775.

It adds the `session.max_shape_override` session option:

```text
input_ids:[8,4096];attention_mask:[8,4096]
```

ORT validates the configured graph-input shapes before cloning, applies them to a disposable shadow graph, runs recursive shape inference, and captures concrete intermediate `NodeArg` shapes without changing runtime input constraints or executable-graph metadata.

The resolved shapes are consumed at:

- **Level 1:** CUDA `MatMulNBits` workspace estimation during `GetCapability()`.
- **Level 2:** `DeclareWorkspaceRequirements()` on constructed kernels during `FinalizeSessionState()`.

Static graph shapes remain usable at Level 2 without configuring an override. Nodes whose inputs remain unresolved retain the existing dynamic allocation/estimation fallback.

The shadow workflow:

- Shares existing initializer `OrtValue`s instead of copying large weights.
- Preserves model-local functions and dynamically generated schemas.
- Captures main-graph and control-flow-subgraph shapes under separate graph identities.
- Matches source and shadow control-flow nodes by stable identity instead of topological position.
- Reuses results when the graph is unchanged and rebuilds them after layout, fusion, or partitioning mutations.
- Converts shadow-model construction and resolution exceptions to `Status`.

Invalid overrides fail with explicit errors, including unknown inputs, rank mismatches, static-dimension conflicts, inconsistent symbolic dimensions, empty entries, and malformed dimension lists. Minimal builds warn and ignore the option rather than failing session initialization.

The inferred downstream shapes are estimation hints, not guaranteed upper bounds: operator shape transformations are not necessarily monotonic. Future preallocation must retain runtime bounds checks and allocation fallbacks.

Level-2 declarations are currently observational. Workspace offset planning and preallocation remain follow-up work.

### Motivation and Context

Dynamic transformer models commonly retain symbolic batch and sequence dimensions, preventing workspace requirements from being calculated before execution.

Deployment shape hints make partition-time budgeting and post-kernel workspace declaration possible while preserving dynamic runtime behavior. The isolated shadow graph prevents optimizers and runtime validation from treating those hints as the model's static shape contract.

This is part of the constrained-environment memory roadmap tracked by #29775 and documented in `docs/annotated_partitioning/future_directions_constrained_env.md`.

Kernels that do not implement `DeclareWorkspaceRequirements()` retain the default no-op and continue using their existing scratch-allocation path.

### Testing

- Built the Debug `onnxruntime_test_all` target on Windows.
- Ran all 20 `MaxShapeOverride.*` tests.
- Added coverage for parsing, malformed separators, zero extents, shape propagation, canonical-graph isolation, invalid overrides, and stable matching of parallel control-flow subgraphs.

### CODE REVIEW

**Comment by titaiwangms** [comment]
## Review (automated multi-agent pass)

Cross-checked against issue #29775, PR #29811, and `docs/annotated_partitioning/*`. Five reviewers examined the diff in parallel for readability, correctness/idiom, adversarial edge cases, spec/invariant fidelity, and cross-module integration.

### 🔴 Critical
**Positional topo-order pairing in `CaptureGraphShapes` is unsound** (`onnxruntime/core/framework/max_shape_override.cc:75-100`). It pairs source/shadow nodes by index in `GetNodesInTopologicalOrder()` and only checks `OpType`/`Domain` for equality. But ORT's topological sort is index-order-dependent (`Graph::PerformTopologicalSortAndCheckIsAcyclic`), and the shadow graph is serialized in the *source's* topo order then re-resolved from scratch — its own topo order is not guaranteed to reproduce the same sequence for independent sibling branches. This was verified with a concrete simulation on a minimal 2-branch graph (`A1→A2`, `B1→B2` in definition order `A1,B1,A2,B2`): source topo order comes out `[A1, B1, B2, A2]` while the shadow's comes out `[A1, B1, A2, B2]` — a mismatch.

Consequences on ordinary branched models (two parallel Convs/Ifs, residual blocks):
1. If the swapped nodes have different `OpType`/`Domain`, the `ORT_RETURN_IF` check fires and `session.max_shape_override` causes an outright session-creation failure on what's meant to be a diagnostic-only feature.
2. If the op types happen to coincide, the check silently passes and the subgraph recursion pairs the **wrong** subgraphs, silently attributing shadow-inferred shapes to the wrong source `Graph` identity for both Level-1 and Level-2.

No existing test exercises a control-flow subgraph or a graph with parallel/independent nodes, so this ships uncaught (`max_shape_override_test.cc` only builds a single-`Identity` chain, which is trivially a fixed point of any topo sort).

**Suggested fix:** pair nodes by a stable identity (e.g. look the shadow node up by `Node::Name()`, which round-trips through `ToProto`) instead of position; keep the list-size check as a sanity assert.

### 🟠 Major
- **Level-2 loop in `session_state.cc` runs unconditionally**, even when `max_shape_override` is unset. `ResolveNodeInputShapes` falls back to static graph shapes, so `DeclareWorkspaceRequirements` is now invoked on every kernel of every session, everywhere — not just when the feature is enabled. Currently benign (no kernel besides `MatMulNBits`/CUDA overrides it, and the default is a no-op), but `ORT_RETURN_IF_ERROR` would turn a future overriding kernel's error into a hard session-init failure on a path the PR describes as log-only, and it adds an unconditional O(#nodes) pass to `FinalizeSessionState` for all users. Suggest gating on `!max_shape_inference_result.Empty()`.
- **Minimal-build hard failure**: setting `session.max_shape_override` on a minimal build makes `InferMaxShapes` return `NOT_IMPLEMENTED`, which `ORT_RETURN_IF_ERROR` turns into a session-load abort rather than a graceful no-op/warning.
- **"Maximum" shapes aren't provably conservative.** Propagating an override value to sibling inputs sharing a symbolic dim is spec-correct (ONNX IR.md confirms dimension variables are unscoped/shared graph-wide), but the shadow-inferred shape for downstream NodeArgs is only a true upper bound if every op on the path is shape-monotonic in that dimension — not guaranteed for `Slice`/`Split`/computed-index `Gather` patterns. This is fine while results stay log-only, but is a real risk if a future PR (per the roadmap's Phase A/B) starts trusting these numbers to skip arena allocation without a runtime bounds-check. Worth stating explicitly in the `kOrtSessionOptionsMaxShapeOverride` doc comment.
- **Cost multiplication with no caching**: the full shadow-graph clone (`ModelProto` serialize → `Model` construct → `Graph::Resolve()`, i.e. full ONNX shape inference) runs independently once per EP in `graph_partitioner.cc`, again after any layout-transform pass that modifies the graph, and again from scratch in `session_state.cc`'s `FinalizeSessionState`. Also, invalid overrides (unknown input, rank mismatch, etc.) are only validated *after* the expensive clone/schema-collection work, so a malformed config on a large model pays the full cost before failing.
- The design doc's own "Shape source is not yet wired" section (`docs/annotated_partitioning/future_directions_constrained_env.md`) is now stale, since this PR closes that gap — worth a follow-up doc update.

### 🟡 Minor
- `node_shape_resolver.h`'s static-shape fallback rejects `dim_value() == 0` as "unresolvable" — 0 is a valid concrete tensor extent and shouldn't force the dynamic-fallback path.
- `ParseMaxShapeOverride` silently accepts a trailing comma (`name:[1,]` parses as `[1]`) and mis-messages on a stray `;;` (empty entry check doesn't fire because the leading delimiter is included in the substring).
- No try/catch around `Model`/`Graph::Resolve()` construction inside the `Status`-returning `InferMaxShapes`, unlike the established `Model::Load` convention of catching and converting exceptions — a malformed model/override could let a C++ exception escape a `Status` boundary in `graph_partitioner.cc`/`session_state.cc`.
- Level-2 shape propagation silently stops at dynamically-generated fused-node schemas (post-partition graph), degrading exactly the CNN heavy-fusion case the design doc calls out ("Level 2 matters more for CNN models with heavy fusion") — worth documenting as a known limitation.
- `const void*` graph-identity keying in `MaxShapeInferenceResult` is a theoretical ABA hazard (pointer reuse after a subgraph is destroyed/replaced) — likely fine given the current compute-then-immediately-consume lifecycle, but fragile as an API; consider a typed `const Graph*` key instead, since a forward declaration is already used elsewhere in this PR.
- Readability: the `max_shape_override.h` vs. `max_shape_inference.h` split is confusing (names differ only by one word, and the more central `InferMaxShapes` lives in the less-obvious header); the new ~28-line Level-2 wiring block in `session_state.cc` is a bare unnamed scope and would read better as a named helper method; `ParseMaxShapeOverride` and `CaptureGraphShapes` each handle more than one concern per function.

### ✅ Verified non-issues
- `LeadingDimProduct` vs. the pre-existing `StaticLeadingDimProduct`: confirmed bit-for-bit consistent on all reachable inputs (including rank-0/rank-1 and `SafeInt` overflow fallback), not a divergence.
- `alignment_bytes`'s "unused by any kernel today" contract from #29811 still holds — nothing in this PR reads or ascribes new meaning to it.
- Sibling-input symbolic-dim propagation is spec-mandated by ONNX IR.md, not a heuristic guess.
- The PR's own stated scope (Level-1 log-only, Level-2 log-only, no behavior change when the option is unset) holds except for the unconditional Level-2 loop noted above.

---
*Generated by an automated multi-agent review pass (readability / correctness / adversarial / spec / integration reviewers).*

**Comment by titaiwangms** [comment]
## Follow-up: verified fixes in the latest push

Re-checked the 5 new commits (2026-08-07) against my earlier review. Confirming fixes:

### 🔴 Critical — **Fixed**
The topo-order positional pairing bug is resolved. `CaptureGraphShapes` (`max_shape_override.cc`) now matches source/shadow nodes-with-subgraphs by a stable identity (`GetSubgraphNodeKey`: node name, falling back to first output name) built into a hash map, rather than by lockstep position in `GetNodesInTopologicalOrder()`. It also now explicitly detects and errors on unmatched/duplicate shadow nodes. The new test `InferMatchesParallelControlFlowSubgraphsByStableIdentity` adds exactly the missing coverage — two independent parallel `If` nodes each with their own subgraph and distinct override shapes, verifying no cross-pairing occurs. This closes both the correctness bug and the test-coverage gap.

### 🟠 Major — mostly addressed
- **Minimal-build hard failure**: fixed. `InferMaxShapes` now returns `Status::OK()` (no-op) under `ORT_MINIMAL_BUILD` instead of `NOT_IMPLEMENTED`, and `session_state.cc` logs a `WARNING` and skips instead of propagating an error.
- **Validation before expensive clone**: fixed. `ValidateMaxShapeOverrides` now runs first (name/rank/static-conflict/symbolic-consistency checks) before any `ModelProto` serialization or shadow `Model` construction.
- **Uncaught exceptions crossing the `Status` boundary**: fixed. Shadow `Model` construction and `Graph::Resolve()` are now wrapped in `ConvertShadowInferenceExceptionsToStatus`, converting `OnnxRuntimeException`/`std::exception` into a `Status`.
- **Parser edge cases** (trailing/leading/double commas, empty entries via stray `;;`): fixed — explicit `ORT_RETURN_IF` checks for empty dimension tokens and empty entries, with new tests `ErrorEmptyDimension` / `ErrorEmptyEntry` covering exactly these cases.
- **`node_shape_resolver.h` zero-extent bug**: fixed — condition changed from `dim_value() > 0` to `dim_value() >= 0`, with a new regression test (`ResolveNodeInputShapesAcceptsStaticZeroExtent`).
- **"Maximum shape is provably conservative" concern**: addressed at the documentation level — `future_directions_constrained_env.md` now explicitly states shape-inference results from the override are "an estimation hint, not a guaranteed upper bound" (citing non-monotonic ops like negative-index `Slice`) and that "pre-allocation must retain runtime bounds checks and an allocation fallback." Good, this sets the right expectation for whoever builds Phase A/B on top of this.
- **Stale doc section**: fixed — "Shape source is not yet wired" section rewritten to "Shape source wiring," accurately describing the new two-source resolution (static graph shapes vs. shadow-graph hints) and explicitly justifying why Level 1 and Level 2 use separately-computed results rather than sharing one (fusion/generated-schema/subgraph-identity changes between partition-time and finalize-time graphs).

### 🟡 Minor — not addressed (still open, low priority)
- Level-2 loop in `session_state.cc` still runs unconditionally regardless of whether `max_shape_override` is set (relies on the default no-op `DeclareWorkspaceRequirements()` being safe for all kernels today). The new doc text confirms this is intentional design, not an oversight, so I'm downgrading this from a blocking concern to a documented tradeoff — worth keeping an eye on if a future kernel's override can fail non-trivially.
- Cost-multiplication / no caching across the 3 shadow-inference call sites: not changed. The doc now explains *why* caching isn't attempted ("deeper caching requires a reliable graph-mutation generation rather than assuming equal topology"), which is a reasonable justification for deferring it, but the repeated full-model shape-inference cost itself remains.
- `const void*` graph-identity keying: unchanged (still a raw pointer key, still fine given the compute-then-consume lifecycle).

**Net assessment:** the Critical soundness bug and all Major correctness/safety gaps are fixed with matching regression tests; remaining items are performance/design tradeoffs that are now explicitly documented rather than silent. This is in much better shape for merge.

**Review by titaiwangms — APPROVED** [review]
Thanks!
